tools/state-parts-dump-check/src/cli.rs

```diff
@@ -555,7 +565,106 @@ async fn run_single_check(
 
     let duration = start.elapsed();
     tracing::info!("Time elapsed in downloading and validating the parts is: {:?}", duration);
-    Ok(StatePartsDumpCheckStatus::Done { epoch_height: epoch_height })
+    Ok(true)
+}
+
+// download and validate state headers for a single epoch and shard
+async fn check_headers(
+    chain_id: &String,
+    epoch_id: &EpochId,
+    epoch_height: u64,
+    shard_id: ShardId,
+    external: &ExternalConnection,
+) -> anyhow::Result<bool> {
+    let directory_path = external_storage_location_directory(
+        &chain_id,
+        &epoch_id,
+        epoch_height,
+        shard_id,
+        &StateFileType::StateHeader,
+    );
+    tracing::info!(directory_path, "the storage location for the state header being checked:");
+    if !external
+        .is_state_sync_header_stored_for_epoch(shard_id, chain_id, epoch_id, epoch_height)
+        .await?
+    {
+        tracing::info!(epoch_height, shard_id, "Waiting for header to be dumped.");
+        return Ok(false);
+    }
+
+    crate::metrics::STATE_SYNC_DUMP_CHECK_NUM_HEADERS_DUMPED
+        .with_label_values(&[&shard_id.to_string(), &chain_id.to_string()])
+        .set(1 as i64);
+
+    tracing::info!(shard_id, epoch_height, "Download and validate state header.");
+
+    let start = Instant::now();
+    let chain_id = chain_id.clone();
+    let epoch_id = epoch_id.clone();
+    let external = external.clone();
+
+    process_header_with_3_retries(chain_id, epoch_id, epoch_height, shard_id, external)
+        .await
+        .unwrap();
```

Comment:
hmm this `unwrap()` looks kind of suspect... Since it is very possible that we get an error here if something is wrong with the external storage. In that case it seems to me that it's better to just log the error and continue instead of panicking right? Same goes for the end of the `check_parts()` function. But since you're not adding that in this PR I guess we can leave it, but it doesn't look right to me

Reply:
Right, we will use `?` instead of `unwrap()` to propagate the error. I have changed it in both parts.